lib/loglevel.js

```diff
@@ -204,37 +219,29 @@
       self.methodFactory = factory || defaultMethodFactory;
 
       self.getLevel = function () {
-          return currentLevel;
+          return currentLevel == null ? defaultLevel : currentLevel;
       };
 
       self.setLevel = function (level, persist) {
-          if (typeof level === "string" && self.levels[level.toUpperCase()] !== undefined) {
-              level = self.levels[level.toUpperCase()];
-          }
-          if (typeof level === "number" && level >= 0 && level <= self.levels.SILENT) {
-              currentLevel = level;
-              if (persist !== false) {  // defaults to true
-                  persistLevelIfPossible(level);
-              }
-              replaceLoggingMethods.call(self, level, name);
-              if (typeof console === undefinedType && level < self.levels.SILENT) {
-                  return "No console available for logging";
-              }
-          } else {
-              throw "log.setLevel() called with invalid level: " + level;
+          level = normalizeLevel(level);
+          currentLevel = level;
+          if (persist !== false) {  // defaults to true
+              persistLevelIfPossible(level);
           }
+          return self.rebuild();
       };
 
       self.setDefaultLevel = function (level) {
-          defaultLevel = level;
-          if (!getPersistedLevel()) {
-              self.setLevel(level, false);
+          defaultLevel = normalizeLevel(level);
+          if (currentLevel == null) {
+            self.rebuild();
```

Comment:
I'm not clear why the defaultlevel logic everywhere needs to change to implement this, can you clarify that? It seems without changing this, rebuild should be able to just call `setDefaultLevel` on children and they would update as expected already.

Reply:
Hmmm, I’m not totally sure what you are getting at here, can you clarify a bit? (What do you mean by “the defaulLevel logic everywhere?”)

The problem I encountered going into this is that, under the hood, `setDefaultLevel()` calls `setLevel()`, and `currentLevel` is whatever level the logger is at. BUT that means you can’t tell the difference between a case where the logger is taking its level from its parent vs. a case where a user has explicitly set the logger’s level. There were three solutions that I could think of for that:

1. Have `currentLevel` be the explicitly set level (or null/undefined if not set delegating to the parent’s level) and have `defaultLevel` represent the parent’s level (since we don’t have a reference to the parent). This is what I did because it seemed most straightforward to *me*, logic-wise.

2. Have a new private variable named `hasExplicitLevel` or similar that indicates whether `currentLevel` comes from a user’s call to `setLevel()` (if true) vs. if it came from the parent logger (if false).

3. Have a logger keep a reference to its parent and drop the `defaultLevel` private variable. This has an issue where `getLevel()` might return something that doesn’t match the actual current behavior (if a parent’s level has changed but this logger has not been rebuilt).

(That said, I suspect most of the above is probably clear to you, and I’m missing what you’re getting at.)

Reply:
I think this should also applies to `setDefaultLevel` though.

Reading the code, I think the current model implemented there is that 'something persisted' means that the value was overridden (and so shouldn't be changed) and anything else means that the value is a default value which should change when the default level changes.

We shouldn't change that existing logic for default levels (whether or not it's a good design, it's a breaking change that would unexpectedly change what's logged by people's applications).

I think the same logic should simply apply for the new rebuild API: if the pre-existing setDefaultLevel logic would modify the level, then rebuilding should modify the level.

In a future version we could re-evaulate this, and in that case I think there's a good argument for explicitly tracking whether the level was set manually or not, but for now I think we have to stick with the existing behaviour.

Reply:
> Reading the code, I think the current model implemented there is that 'something persisted' means that the value was overridden (and so shouldn't be changed) and anything else means that the value is a default value which should change when the default level changes.

I guess I’d argue that leaning on persistence is already problematic, since it doesn’t work in environments that have no persistence mechanism (e.g. Node, Bun, an `iframe` with a lot of sandboxing options turned on, etc.). It means this happens, which sure feels like a bug to me:

```js
var child = log.getLogger("test");
child.getLevel();  // 3 (WARN)

child.setLevel("TRACE");
child.getLevel();  // 0 (TRACE), good!

child.setDefaultLevel("DEBUG");
child.getLevel();  // 1 (DEBUG), seems bad!
```

Or more saliently, once there’s a `rebuild()` method:

```js
log.getLevel();  // 3 (WARN)
var child = log.getLogger("test");
child.getLevel();  // 3 (WARN), it inherited, good.

child.setLevel("TRACE");
child.getLevel();  // 0 (TRACE), good!

log.rebuild();
child.getLevel();  // 3 (WARN), seems bad that rebuilding changes the
                    // child's level after you've explicitly set it.
```

I suppose this is, very subtly, breaking, the question is whether the current behavior is a bug (my view, and hence why it is OK to “fix” as a byproduct of this) or is just what it is. I think most people’s mental model here is what I’m doing in this change and so wouldn’t be surprised by it.

Reply:
There is also a test that models the above scenario: https://github.com/pimterry/loglevel/blob/45396397b87dacc20a62bc9bf303a70c61854be1/test/multiple-logger-test.js#L175-L182

I feel like it is clear to me that that test should always succeed (and therefore needs this change, or one of the similar things I enumerated above).

Reply:
Ugh, looked back at the discussion when I originally added `setDefaultLevel()`, and I am definitely wrong. “Default’ was probably a confusing name choice (my fault!) and I was over-focused on the browser context at the time. Anyway, I had lost track of the fact that `setDefaultLevel()` was originally intended to be a way to say “set the level unless you’ve loaded saved configuration info from somewhere,” *not* “set the level if one hasn’t been explicitly set otherwise.” The idea was that `setDefaultLevel()` was for built-in config in your app, and `setLevel()` was only for runtime *re*configuration by users (e.g. a dev setting it in their console, some preference a user might adjust, a CLI flag).

Obviously I need to re-do this. `setDefaultLevel()` might not have been well designed (my fault!), but it is working as designed and is far too well baked to change, as you noted.

> I think the same logic should simply apply for the new rebuild API: if the pre-existing setDefaultLevel logic would modify the level, then rebuilding should modify the level.

I don’t think this is quite right, though. With the right definition for `setDefaultLevel()` (above), it seems like it obviously qualifies as explicitly setting the logger’s level just as much as `setLevel()` does. Using either one means the logger’s level shouldn’t be influenced by its parent anymore.

That is:

```js
log.setLevel("WARN");

var child1 = log.getLogger("child1");
child1.setLevel("ERROR");
child1.getLevel();  // ERROR

var child2 = log.getLogger("child2");
child2.setDefaultLevel("ERROR");
child2.getLevel();  // Let's assume ERROR and something was not previously saved

var child3 = log.getLogger("child3");
child3.getLevel():  // Let's assume it WARN because nothing was previously saved

log.methodFactory = function() { ... }
log.setLevel("TRACE");
log.rebuild();

// So the levels should now be:
//   log    = TRACE (we just set it)
//   child1 = ERROR (was set with setLevel)
//   child2 = ERROR (was set with setDefaultLevel)
//   child3 = TRACE (was never set, so inherits the change from the parent)
```

Let me know if you think the above doesn’t make sense, but it seems right to me. I think the disconnect here is over what level `child2` should be at.

This also makes it a bit ambiguous what `resetLevel()` should do. As coded today, calling it on the children would change `child1` to TRACE and `child2` would remain at `ERROR`. I think that’s probably fine — it resets from the *persisted* level to the logger’s expected level — but I can imagine people expecting it clears any level info on the logger in favor of inheriting from the parent instead.

Reply:
OK, this should be solved. I kept getting myself twisted up in the logic with the boolean, so I wound up refactoring this as 3 private variables. LMK know if you find this too confusing and I will refactor some more, but it felt much easier to comprehend for me:

- `inheritedLevel` is a cache of parent logger's level (or `WARN` for the root logger). This is always set to something valid.
- `defaultLevel` is the default level that has been set for this logger. It may be null (unset).
- `userLevel` is the level that has been set for this logger via persistence or via `setLevel()`. It may be null (unset).

The current/effective level of the logger is `userLevel OR defaultLevel OR inheritedLevel`.

This seemed like a more straightforward model of what’s going on, and the conceptual order of precedence is relatively obvious. There's no boolean or other state that is modifying how the levels behave based on what methods have been called in what contexts (`resetLevel()` was causing me subtle problems, although after the above renaming, I am pretty sure I could switch back to having the boolean and get it right this time).

---

The other thing worth noting here is that I opted to have child loggers reach up to their parent to get the parent’s level, where I previously had the parent updating the children’s levels. Based on the discussion here, we would have had to add a special public method for the parent to call, which seemed bad. So I flipped it around.

(Side note: since we only have one level of children right now, the children can just grab the parent level from `defaultLogger`. In some hypothetical future with a deep hierarchy of loggers, each logger will need to hold a reference to its parent.)